Mathlib/Tactic/Bound.lean

```diff
@@ -0,0 +1,273 @@
+
+-- <
+attribute [bound] Nat.cast_pos_of_pos neg_lt_neg sub_lt_sub_left sub_lt_sub_right add_lt_add_left
+  add_lt_add_right mul_lt_mul_left_of_pos_of_lt mul_lt_mul_right_of_pos_of_lt div_lt_div_of_pos_left
+  div_lt_div_of_pos_right pow_lt_pow_left div_lt_self one_lt_div_of_pos_of_lt
+  div_lt_one_of_pos_of_lt
+
+-- min and max
+attribute [bound] min_le_right min_le_left le_max_left le_max_right le_min max_le lt_min max_lt
+
+-- Memorize a few constants to avoid going to `norm_num`
+attribute [bound] zero_le_one zero_lt_one zero_le_two zero_lt_two
+
+/-!
+### Forward rules for `bound`
+-/
+
+-- Bound applies `le_of_lt` to all hypotheses
+attribute [bound_forward] le_of_lt
+
+/-!
+### Guessing rules: when we don't know how to recurse
+-/
+
+section Guessing
+
+variable {α : Type} [LinearOrder α] {a b c : α}
+
+-- `min` and `max` guessing lemmas
+lemma le_max_of_le_left_or_le_right : a ≤ b ∨ a ≤ c → a ≤ max b c := le_max_iff.mpr
+lemma lt_max_of_lt_left_or_lt_right : a < b ∨ a < c → a < max b c := lt_max_iff.mpr
+lemma min_le_of_left_le_or_right_le : a ≤ c ∨ b ≤ c → min a b ≤ c := min_le_iff.mpr
+lemma min_lt_of_left_lt_or_right_lt : a < c ∨ b < c → min a b < c := min_lt_iff.mpr
+
+/-- Branch on `1 ≤ a ∨ a ≤ 1` for `a^n` -/
+lemma pow_le_pow_right_of_le_one_or_one_le {R : Type} [OrderedSemiring R] {a : R} {n m : ℕ}
+    (h : 1 ≤ a ∧ n ≤ m ∨ 0 ≤ a ∧ a ≤ 1 ∧ m ≤ n) : a ^ n ≤ a ^ m := by
+  rcases h with ⟨a1, nm⟩ | ⟨a0, a1, mn⟩
+  · exact pow_le_pow_right a1 nm
+  · exact pow_le_pow_of_le_one a0 a1 mn
+
+-- Register guessing rules
+attribute [bound]
+  -- Which side of the `max` should we use as the lower bound?
+  le_max_of_le_left_or_le_right
+  lt_max_of_lt_left_or_lt_right
+  -- Which side of the `min` should we use as the upper bound?
+  min_le_of_left_le_or_right_le
+  min_lt_of_left_lt_or_right_lt
+  -- Given `a^m ≤ a^n`, is `1 ≤ a` or `a ≤ 1`?
+  pow_le_pow_right_of_le_one_or_one_le
+
+end Guessing
+
+/-!
+### Closing tactics
+-/
+
+/-- Close numerical goals with `norm_num` -/
+def boundNormNum : Aesop.RuleTac :=
+  Aesop.SingleRuleTac.toRuleTac fun i => do
+    let tac := do Mathlib.Meta.NormNum.elabNormNum .missing .missing .missing
+    let goals ← Lean.Elab.Tactic.run i.goal tac |>.run'
+    if !goals.isEmpty then failure
+    return (#[], none, some .hundred)
+attribute [aesop unsafe 10% tactic (rule_sets := [Bound])] boundNormNum
+
+/-- Close numerical and other goals with `linarith` -/
+def boundLinarith : Aesop.RuleTac :=
+  Aesop.SingleRuleTac.toRuleTac fun i => do
+    Linarith.linarith false [] {} i.goal
+    return (#[], none, some .hundred)
+attribute [aesop unsafe 5% tactic (rule_sets := [Bound])] boundLinarith
+
+/-!
+### `bound` tactic implementation
+-/
+
+/-- Pull the array out of an optional `"[" term,* "]"` syntax, or return `#[]` -/
+def maybeTerms : Syntax → Array Syntax
+  | node _ _ #[_, node _ _ s, _] => s.getEvenElems
+  | _ => #[]
+
+/-- Add extra hypotheses -/
```

Comment:
Deserves more explanation

Reply:
Added more; is that enough?

Reply:
Looks good, and makes it clear that it too is not `bound`-specific